packages/melody-parser/src/util.js

```diff
@@ -44,6 +44,16 @@ export function copyEnd(node, end) {
     return node;
 }
 
+export function copySource(node, entireSource) {
+    if (entireSource && node.loc.start && node.loc.end) {
```

Comment:
Instead of eagerly creating a lot of substrings (potential performance and memory impact) you might be able to save the full source string and only do this computation when reading the property.

Reply:
Yes, that would be a lot nicer. My initial thought is to save the full source string on the AST root.

There might be a small problem with expressions and Twig tags because the starting positions often don't match what I need for pretty printing purposes, but it should be possible. I might have to add one or two additional fields per node, but be able to drop the `originalSource` field.